### PR TITLE
extend SMC CLI for reencryption

### PR DESCRIPTION
**Pull request by jbsv**

### CODE REVIEW

File: server/scripts/test_reveal.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
```
Comment (pierluca) [inline server/scripts/test_reveal.sh:2]: I seem to recall you wanted to check that parameter $1 was correctly provided here.

File: server/smc/smccli/controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+// Build implements node.Initializer. In this case we don't need any command.
+func (s smcctl) SetCommands(builder node.Builder) {
+	cmd := builder.SetCommand("smc")
+	cmd.SetDescription("SMC service administration")
+
+	sub := cmd.SetSubCommand("createkeys")
+	sub.SetDescription("create key pair for reencryption")
+	sub.SetAction(builder.MakeAction(createKpAction{}))
```
Comment (pierluca) [inline server/smc/smccli/controller/controller.go:32]: as we discussed, this should be a local action. No need for `builder.MakeAction`, you can just use pure functions.

File: server/smc/smccli/controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		cli.StringFlag{
+			Name:  "encrypted",
+			Usage: "the encrypted string, as <hex(K)>:<hex(C1):<hex(C2):...>",
+		},
+		cli.StringFlag{
+			Name:  "privk",
+			Usage: "drop me if you can",
+		},
+	)
+	sub.SetAction(builder.MakeAction(revealAction{}))
```
Comment (pierluca) [inline server/smc/smccli/controller/controller.go:54]: same here

File: server/smc/smccli/smccli.go [hunk 4]
```diff
@@ -2,14 +2,13 @@ package main
...
-	"os"
-
 	"go.dedis.ch/dela/cli/node"
-
 	db "go.dedis.ch/dela/core/store/kv/controller"
 	dkg "go.dedis.ch/dela/dkg/pedersen/controller"
 	minogrpc "go.dedis.ch/dela/mino/minogrpc/controller"
+	smcctl "go.dedis.ch/hbt/smc/smccli/controller"
+	"io"
+	"os"
```
Comment (pierluca) [inline server/smc/smccli/smccli.go:11]: why change the import order ?

Normally we use the goimports convention, standard packages first.

File: server/smc/smccli/controller/reveal.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	dela.Logger.Debug().Msgf("xc + xcInv: %v", sum)
+
+	XhatDec := suite.Point().Mul(xcInv, dkgPk)
+	dela.Logger.Debug().Msgf("XhatDec:%v", XhatDec)
+
+	Xhat := suite.Point().Add(XhatEnc, XhatDec)
+	dela.Logger.Debug().Msgf("Xhat:%v", Xhat)
+
+	XhatInv := suite.Point().Neg(Xhat)
+	dela.Logger.Debug().Msgf("XhatInv:%v", XhatInv)
```
Comment (pierluca) [inline server/smc/smccli/controller/reveal.go:28]: Do all these logs actually bring any value ? They read like debugging printfs.

File: server/smc/smccli/controller/reveal.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+	// Decrypt Cs to keyPointHat
+	for _, C := range Cs {
+		dela.Logger.Debug().Msgf("C:%v", C)
+
+		keyPointHat := suite.Point().Add(C, XhatInv)
+		dela.Logger.Debug().Msgf("keyPointHat:%v", keyPointHat)
+
+		keyPart, err := keyPointHat.Data()
+		dela.Logger.Debug().Msgf("keyPart:%v", keyPart)
```
Comment (pierluca) [inline server/smc/smccli/controller/reveal.go:40]: same for these ones

File: server/smc/smccli/controller/reveal.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+		if err != nil {
+			e := xerrors.Errorf("Error while decrypting Cs: %v", err)
+			dela.Logger.Error().Msg(e.Error())
```
Comment (pierluca) [inline server/smc/smccli/controller/reveal.go:44]:
```suggestion
			dela.Logger.Error().Err(e).Msg("Failed revealing secret")
```

File: server/smc/smccli/controller/controller.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
+package controller
+
+import (
+	"go.dedis.ch/dela/cli"
+	"go.dedis.ch/dela/cli/node"
+)
+
+// smcctl is an initializer with a set of commands for the SMC. It only
+// creates and injects a new SMC instance.
```
Comment (pierluca) [inline server/smc/smccli/controller/controller.go:9]: does it inject a `SMC instance` ? It just configures the CLI for the SMC , no ?

File: server/smc/smccli/controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+//
+// - implements node.Initializer
+type smcctl struct{}
+
+// NewSmcController returns a new SMC initializer
+func NewSmcController() node.Initializer {
+	return smcctl{}
+}
+
+// Build implements node.Initializer. In this case we don't need any command.
```
Comment (pierluca) [inline server/smc/smccli/controller/controller.go:19]: This comment, or the function name, is out of date (I don't remember the interface's expectations)

**Review by pierluca — COMMENTED** [review]
The actions were not reviewed as they need to be refactored.

File: server/smc/smccli/smccli.go (hunk 4 above)
Comment (jbsv) [reply server/smc/smccli/smccli.go:11]: don't know: I see that Goland auto-magically re-order things, which is a PITT

File: server/scripts/test_reveal.sh (hunk 1 above)
Comment (jbsv) [reply server/scripts/test_reveal.sh:2]: done

File: server/smc/smccli/controller/controller.go (hunk 3 above)
Comment (jbsv) [reply server/smc/smccli/controller/controller.go]: done

File: server/smc/smccli/controller/controller.go (hunk 2 above)
Comment (jbsv) [reply server/smc/smccli/controller/controller.go]: done

File: server/smc/smccli/controller/reveal.go (hunk 5 above)
Comment (jbsv) [reply server/smc/smccli/controller/reveal.go]: cleaned up

File: server/smc/smccli/controller/reveal.go (hunk 7 above)
Comment (jbsv) [reply server/smc/smccli/controller/reveal.go]: cleaned up

File: server/smc/smccli/smccli.go (hunk 4 above)
Comment (jbsv) [reply server/smc/smccli/smccli.go]: done

**Review by pierluca — APPROVED** [review]

**Pull request merged**